### PR TITLE
GoSYSL

### PR DESCRIPTION
**Pull request by sahejsingh**
### Introducing GoSYSL
SYSL implemented in GO-lang using parser & lexer generated by antlr. 

( Apologies in advance as this is big pull request. See below on where to start)

### Main Features
Implements all of the following sysl language features:
- Define Application and Endpoints
- Define multi-level, Rest endpoints
- Recognises all of the HTTP verbs
- Collector syntax (for projects)
- Define parameters to endpoints and pass arguments
- Tuple and Relational types
- Recognises all the sysl supported data types like int, int32, int64, string, date, datetime etc. 
- In place definition of types
- Events (pub-sub)
- Mixins
- Looping constructs like For, IF, Else, Loop, While, One of, Loop N, Alt, Until.
- Generates text and json protobuf output. The json output is used by vscode-sysl extension to provide language support in Visual Studio Code.
 
The grammar diverges very slightly from the original version. It does not allow for "." in the Application and Endpoint names.

### Known Limitations

- Grammar does not handle transformation expressions yet.

### Tests
- See tests/*.sysl
- These tests compares the textpb output generated by legacy python with the output generated by gosysl.

To help you get around, here is a bit of description.

The GO code is divided into three parts.
- Generated GO code
  - all in _grammar/_ directory
- Handwritten GO code
  -  _sysl.go_ is the main entry point 
  - _sysl_listener_impl.go_ is the implementation of the ANTLR generated listener interface).
- Tests
  - Parser_test.go

The only other files of interest are the _SyslLexer.g4_ and _SyslParser.g4_.

@anz-bank/sysl-developers

### CODE REVIEW

**Comment by anz-rfc** [comment]
@sahejsingh , apologies for this fairly superficial/meta comment, but i guess my first reaction is that a merge request with a +24k line delta probably isnt going to be reviewable (especially not through the github PR interface)

Is a lot of the size of this change due to e.g. generated go code?

Are there any relatively low-effort moves you could do to break this into multiple independent PRs or otherwise make it easier to review?

**Comment by sahejsingh** [comment]
@anz-rfc The majority of the code is generated. See the part in my comment in the Pull request above where I have described where to start.

Copied here again....

To help you get around, here is a bit of description.

The GO code is divided into three parts.

- Generated GO code
  - all in grammar/ directory
- Handwritten GO code
  - sysl.go is the main entry point
  - sysl_listener_impl.go is the implementation of the ANTLR generated listener interface).
- Tests
  - Parser_test.go

The only other files of interest are the _SyslLexer.g4_ and _SyslParser.g4_.

**Comment by anzdaddy** [comment]
Trying to submit a review with 82 comments, I keep getting "405 Not Allowed".

**Comment by anzdaddy** [comment]
Looks like actually worked, just didn't look like it.

File: sysl2/antlr/grammar/SyslLexer.g4 [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+fragment T : [tT];
+fragment U : [uU];
+fragment V : [vV];
+fragment W : [wW];
+fragment X : [xX];
+fragment Y : [yY];
+fragment Z : [zZ];
+
+NativeDataTypes     :
+                    ( (I N T '3' '2') | (I N T '6' '4') | (I N T) | (F L O A T) | ( S T R I N G) | (D A T E) | (B O O L) | (D E C I M A L) | (D A T E T I M E) | (X M L) | (A N Y))
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/grammar/SyslLexer.g4]: nope.
[Ref: Case-Insensitive Lexing](https://github.com/antlr/antlr4/blob/master/doc/case-insensitive-lexing.md)

File: sysl2/antlr/grammar/lexer_impl.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+func calcSpaces(text string) int {
+	s := 0
+
+	for i := 0; i < len(text); i++ {
+		if text[i] == ' ' {
+			s++
+		}
+		if text[i] == '\t' {
+			s += 4
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/grammar/lexer_impl.go]: I am not sure what you mean by 

> Indent rules should not allow tabs and spaces immediately above and below each other

**calcSpaces** is there to handle both tabs and spaces.
Also some of our files had mixed tabs and spaces too. This code handles all of theses cases seamlessly. Makes parser more "fault-tolerant".

File: sysl2/antlr/Parser_test.go [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+		TextPB(m2, "generated.txt")
+		TextPB(&mod, "golden.txt")
+	}
+
+	return result
+}
+
+func TestSimpleEP(t *testing.T) {
+	filename := "tests/test1.sysl"
+	if loadAndCompare(Parse(filename, ""), filename, "") == false {
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/Parser_test.go]: Debugging is easy with separate tests.

[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/Parser_test.go]: IMHO X == false is better than !X, simply because its easier to read.

[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/Parser_test.go]: done what you suggested above.

File: sysl2/antlr/grammar/lexer_impl.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	for k := range keywords {
+		if s.HasPrefix(lower, keywords[k]) {
+			return true
+		}
+	}
+	return false
+}
+
+func createDedentToken(source *antlr.TokenSourceCharStreamPair) *antlr.CommonToken {
+	t := antlr.NewCommonToken(source, SyslLexerDEDENT, 0, 0, 0)
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/grammar/lexer_impl.go]: sure. done.

File: sysl2/antlr/grammar/lexer_impl.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+func createIndentToken(source *antlr.TokenSourceCharStreamPair) *antlr.CommonToken {
+	t := antlr.NewCommonToken(source, SyslLexerINDENT, 0, 0, 0)
+
+	return t
+}
+
+func GettokenName(t antlr.Token, l *SyslLexer) string {
+	if t.GetTokenType() > 0 && t.GetTokenType() < len(l.SymbolicNames) {
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/grammar/lexer_impl.go]: this code will go away, it was added for debug logging. Dont need it anymore,

File: sysl2/antlr/grammar/lexer_impl.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+		return "EOF"
+	}
+
+	return fmt.Sprintf("%d", t.GetTokenType())
+}
+
+var keywords = [...]string{"set of", "return", "for", "one of", "else", "if", "loop", "until", "alt", "while"}
+var prevToken []antlr.Token
+
+type Stack struct {
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/grammar/lexer_impl.go]: Original code was in java where I had the stack class. I didn't want to remove ops on stack.
Also I don't like the way slice operations returns the new object after append. Does not feel like you are working with an object.

File: sysl2/antlr/grammar/lexer_impl.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+		nextTok := prevToken[0]
+		prevToken = prevToken[1:]
+		// tokenName := GettokenName(nextTok, l)
+		return nextTok
+	}
+
+	next := l.BaseLexer.NextToken()
+	// tokenName := GettokenName(next, l)
+	// return NEWLINE
+	if gotNewLine && (next.GetTokenType() == SyslLexerNEWLINE || next.GetTokenType() == SyslLexerNEWLINE_2 || next.GetTokenType() == SyslLexerEMPTY_LINE || next.GetTokenType() == SyslLexerINDENTED_COMMENT || next.GetTokenType() == SyslLexerEMPTY_COMMENT) {
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/grammar/lexer_impl.go]: nice. done.

File: sysl2/antlr/sysl_listener_impl.go [hunk 8]
```diff
@@ -0,0 +1,2225 @@
...
+// ExitInplace_table is called when production inplace_table is exited.
+func (s *TreeShapeListener) ExitInplace_table(ctx *parser.Inplace_tableContext) {}
+
+// EnterTable is called when production table is entered.
+func (s *TreeShapeListener) EnterTable(ctx *parser.TableContext) {
+	if s.typename == "" {
+		s.typename = ctx.Name_str().GetText()
+	} else {
+		s.typename = s.typename + "." + ctx.Name_str().GetText()
+	}
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/sysl_listener_impl.go]: umm, again, I prefer easy to follow logic.

File: sysl2/antlr/sysl.go [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+	case *sysl.Statement_Action:
+		return applied
+	case *sysl.Statement_Ret:
+		return applied
+	default:
+		panic("collector: unhandled type")
+	}
+
+	for _, stmt := range stmts {
+		applied = applyAttributes(src, stmt) || applied
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/sysl.go]: if applied is true, applyAttributes wont be called.

File: sysl2/antlr/sysl.go [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+		}
+	}
+	return true
+}
+
+func isSameCall(a *sysl.Call, b *sysl.Call) bool {
+	return isSameApp(a.Target, b.Target) && a.Endpoint == b.Endpoint
+}
+
+func applyAttributes(src *sysl.Statement, dst *sysl.Statement) bool {
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/sysl.go]: I have added comment in the code.

File: sysl2/antlr/sysl.go [hunk 11]
```diff
@@ -0,0 +1,464 @@
...
+					}
+				}
+			}
+		}
+	}
+	checkEndpointCalls(mod)
+	collectorPubSubCalls(mod)
+}
+
+func fileExists(filename string) bool {
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/sysl.go]: IsNotExist takes the error as the argument.

File: sysl2/antlr/sysl.go [hunk 12]
```diff
@@ -0,0 +1,464 @@
...
+
+	if !fileExists(filename) {
+		if filename[len(filename)-5:] != ".sysl" {
+			filename = filename + ".sysl"
+		}
+		temp := root + "/" + filename
+
+		if !fileExists(temp) {
+			fmt.Fprintln(os.Stderr, fmt.Sprintf("input file does not exist\nFilename: %s\n", temp))
+			os.Exit(ImportError)
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/sysl.go]: Yes. Other tools (like vscode extension) would like to know if the parsing (or other op) was successful.

File: sysl2/antlr/sysl.go [hunk 13]
```diff
@@ -0,0 +1,464 @@
...
+		}
+		filename = temp
+	}
+	var empty struct{}
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/sysl.go]: Ughh...struct{}{}

File: sysl2/antlr/sysl.go [hunk 14]
```diff
@@ -0,0 +1,464 @@
...
+
+		antlr.ParseTreeWalkerDefault.Walk(listener, tree)
+		if len(listener.imports) == 0 {
+			break
+		}
+		imported[filename] = empty
+
+		for len(listener.imports) > 0 {
+			filename = listener.imports[0]
+			listener.imports = listener.imports[1:]
```
[the remark this answers is not in the document]
Comment (sahejsingh) [reply sysl2/antlr/sysl.go]: I am not sure if the order will have impact on the serialised output.

**Pull request merged**